articles/actions.md

```diff
@@ -0,0 +1,437 @@
+There is one action specification per `.action` file.
+
+### Example
+
+```
+# Define a goal of washing all dishes
+bool heavy_duty  # Spend extra time cleaning
+---
+# Define the result that will be published after the action execution ends.
+uint32 total_dishes_cleaned
+---
+# Define a feedback message that will be published during action execution.
+float32 percent_complete
+uint32 number_dishes_cleaned
+```
+
+## Introspection tools
+
+Actions, like topics and services, are introspectable from the command line.
+
+The command line tool, `ros2 action`, will be able to:
+
+- list action names associated with any running action servers or action clients
+- list action servers and action clients
+- display active goals on an action server
+- display the arguments for an action goal
+- display the type of an action's goal, feedback, and result
+- find actions by action type
+- echo feedback, status, and result for an action goal
+- call an action, display feedback as it is received, display the result when received, and cancel the action (when the tool is terminated prematurely).
+
+Each action will be listed and treated as a single unit by this tool.
+
+## Goal States
+
+The action server maintains a state machine for each goal it accepts from a client.
+Rejected goals are not part of the state machine.
+
+![Action Goal State Machine](../img/actions/goal_state_machine.png)
+
+There are three active states:
+
+- **ACCEPTED** - The goal has been accepted and is awaiting execution.
+- **EXECUTING** - The goal is currently being executed by the action server.
+- **CANCELING** - The client has requested that the goal be canceled and the action server has accepted the cancel request.
+This state is useful for any user-defined "clean up" that the action server may have to do.
+
+And three terminal states:
+
+- **SUCCEEDED** - The goal was achieved successfully by the action server.
+- **ABORTED** - The goal was terminated by the action server without an external request.
+- **CANCELED** - The goal was canceled after an external request from an action client.
+
+State transitions triggered by the action server according to its designed behavior:
+
+- **execute** - Start execution of an accepted goal.
+- **set_succeeded** - Notify that the goal completed successfully.
+- **set_aborted** - Notify that an error was encountered during processing of the goal and it had to be aborted.
+- **set_canceled** - Notify that canceling the goal completed successfully.
+
+State transitions triggered by the action client:
+
+- **send_goal** - A goal is sent to the action server.
+The state machine is only started if the action server *accepts* the goal.
+- **cancel_goal** - Request that the action server stop processing the goal.
+A transition only occurs if the action server *accepts* the request to cancel the goal.
+
+## API
+
+Proposed examples can be in the respository [examples](https://github.com/ros2/examples/tree/actions_proposal).
```

Comment:
I think this URL doesn't exist anymore; where are the examples now?

Reply:
Fixed: f4b59fa